Implement UI Toolkit API documentation

andy-armstrong opened the pull request:
## Description

[FEDX-3](https://openedx.atlassian.net/browse/FEDX-3)

This change uses [JSDox](http://jsdox.org/) to automatically generate Markdown files from the UI Toolkit's JavaScript doc strings. I then build a Jekyll-based static site from these raw files, ready to be published to GitHub Pages. Use ``gulp doc`` to generate the documentation and to serve it using browserSync along with file watchers on all of the inputs. It also supports a ``build-preview`` task that will upload the site to S3 for folks to preview.

You can view the generated documentation here: http://ux-test.edx.org/andya/api-doc/

## Manual Test Plan

I've manually verified that each test class is being rendered reasonably, and I've cleaned up the documentation to match.

## Testing Checklist
- [ ] Manually test responsive behavior.

## Post-review
- [ ] Squash commits into discrete sets of changes with descriptive commit messages.

## Reviewers
- [ ] @pdesjardins 
- [x] @bjacobel 

If you've been tagged for review, please check your corresponding box once you've given the :+1:.

We'll also automatically suggest reviewers for this PR based on the code it touches.

FYI @AlasdairSwan @dsjen @dan-f @cahrens

bjacobel commented:
Getting some font loading errors:
```
http://ux-test.edx.org/public/edx-pattern-library/fonts/OpenSans/OpenSans-Bold-webfont.woff2 Failed to load resource: the server responded with a status of 404 (Not Found)
http://ux-test.edx.org/public/edx-pattern-library/fonts/OpenSans/OpenSans-Semibold-webfont.woff2 Failed to load resource: the server responded with a status of 404 (Not Found)
http://ux-test.edx.org/public/edx-pattern-library/fonts/OpenSans/OpenSans-Regular-webfont.woff2 Failed to load resource: the server responded with a status of 404 (Not Found)
http://ux-test.edx.org/public/edx-pattern-library/fonts/OpenSans/OpenSans-Bold-webfont.woff Failed to load resource: the server responded with a status of 404 (Not Found)
http://ux-test.edx.org/public/edx-pattern-library/fonts/OpenSans/OpenSans-Semibold-webfont.woff Failed to load resource: the server responded with a status of 404 (Not Found)
http://ux-test.edx.org/public/edx-pattern-library/fonts/OpenSans/OpenSans-Regular-webfont.woff Failed to load resource: the server responded with a status of 404 (Not Found)
http://ux-test.edx.org/public/edx-pattern-library/fonts/OpenSans/OpenSans-Bold-webfont.ttf Failed to load resource: the server responded with a status of 404 (Not Found)
http://ux-test.edx.org/public/edx-pattern-library/fonts/OpenSans/OpenSans-Semibold-webfont.ttf Failed to load resource: the server responded with a status of 404 (Not Found)
http://ux-test.edx.org/public/edx-pattern-library/fonts/OpenSans/OpenSans-Regular-webfont.ttf Failed to load resource: the server responded with a status of 404 (Not Found)
http://ux-test.edx.org/public/edx-pattern-library/fonts/edx-icons/edx-icons.woff Failed to load resource: the server responded with a status of 404 (Not Found)
http://ux-test.edx.org/public/edx-pattern-library/fonts/edx-icons/edx-icons.ttf Failed to load resource: the server responded with a status of 404 (Not Found)
```

bjacobel commented on README.md:
```diff
@@ -6,6 +6,9 @@ A JavaScript toolkit for building edX user interfaces.
 [![Bower version](https://badge.fury.io/bo/edx-ui-toolkit.svg)](https://badge.fury.io/bo/edx-ui-toolkit)
 [![CoverageStatus](https://img.shields.io/coveralls/edx/edx-ui-toolkit.svg)](https://coveralls.io/r/edx/edx-ui-toolkit?branch=master)
+
+<a href="https://nodei.co/npm/edx-ui-toolkit/"><img src="https://nodei.co/npm/edx-ui-toolkit.png?downloads=true&downloadRank=true&stars=true"></a>
```
Do we need this and also the fury.io badge from line 6?

andy-armstrong replied:
Oh, I forgot I put this in. I was just experimenting to see which was more useful. What do you think? I should move this to a different PR if we want it.

bjacobel replied:
I like it, but I agree about a different PR. Maybe remove the Bower badge at the same time.

bjacobel commented:
The nitpickiest thing: Can we increase the `line-height` a little? (1.6 looked good to me). Makes things a little more readable.

bjacobel commented on gulp/tasks/preview.js:
```diff
@@ -0,0 +1,82 @@
+    var branch = gitUtils.currentBranch();
+    if (previewDomain) {
+        childProcess.execSync(
+            'aws s3 sync ' + previewSiteDir + ' s3://' + previewDomain + '/' + branch
```
This requires you to have `awscli` installed - should either document that somewhere or try to find a Gulp-native way to do it.

andy-armstrong replied:
Good point. I should document how this should work. Here's the steps from the wiki:

https://openedx.atlassian.net/wiki/display/FEDX/How+to+share+static+front+end+changes+for+review

AlasdairSwan commented:
Nice work @andy-armstrong. I noticed an a11y issue in that the contrast ratio between the light blue text and the grey background is only 2.25.

bjacobel commented on gulp/config.js:
```diff
@@ -0,0 +1,51 @@
+            sources: [
+                './src/js/disclosure/*.js',
+                './src/js/dropdown-menu/*.js',
+                './src/js/pagination/*.js'
```
Can we use a path like `'./src/js/**/*.js'` here to avoid having to remember to add new patterns here when we create them?

andy-armstrong replied:
Yep, that would make more sense. It will have to filter out utils so that it doesn't duplicate all of those files.

bjacobel commented on gulp/tasks/preview.js:
```diff
@@ -0,0 +1,82 @@
+        'jekyll-build-preview',
+        'preview-webpack',
+        'upload-preview',
+        callback);
```
`);` should be split from `callback` by a line

andy-armstrong commented:
@bjacobel the font errors are a known problem just with the preview. It is not correctly using the new base URL that the site is being hosted at. I don't think it is a showstopper since it doesn't affect the local Jekyll build, and hence I don't think it will affect GitHub Pages. I will look into it again and see if I can pin it down.

andy-armstrong commented:
Thanks @AlasdairSwan. I was having second thoughts about the gray background, even though that is our edX pattern. Maybe I should do something more like ReadTheDocs:

http://edx-developer-guide.readthedocs.org/en/latest/

@chris-mike what would you recommend?

andy-armstrong commented:
@bjacobel @AlasdairSwan @pdesjardins - thanks for the great feedback. I've fixed the bug with the preview fonts, and I've cleaned up the styling to be cleaner and not have a11y issues. I've pushed up the preview again:

http://ux-test.edx.org/andya/api-doc/utilities/string-utils/

Please re-review.

bjacobel commented:
![syntax highlighting](https://i.bjacobel.com/20160502-zebde.png)

It looks like (something? Jekyll?) is tokenizing and wrapping in `<span>`s the code blocks for syntax highlighting, but our presentation isn't able to do anything with it because it's lacking CSS to style it. Could we look into adding that CSS? I think all we need is [this file](https://github.com/bjacobel/bjacobel.com/blob/master/_sass/_syntax.scss) (but I'm not sure where it comes from).

bjacobel commented:
Some sort of rule along the lines of `pre.highlight {background-color: grey}` might increase the code readability as well.

andy-armstrong commented:
Good point @bjacobel. I intended to look into how the code rendering was meant to work. I'll investigate...

andy-armstrong commented:
@bjacobel I've pushed up changes to support the code styling (I borrowed them from the Pattern Library).

pdesjardins commented:
@andy-armstrong, this looks great. Please merge. What I see at http://ux-test.edx.org/andya/api-doc/ is excellent!

bjacobel commented:
:thumbsup:
